Configure fails to find liblapack unless CC=c++

There is an extern "C" declaration in conftest.c which configure attempts to compile with the C compiler. This correctly fails, yet prevents configure from finding various libraries, e.g. lapack. To reproduce this error, configure with/without setting CC to a C++ compiler:

`./configure --with-lapack=/usr/lib64
`
configure output terminates with:
...

> checking for gawk... gawk
> checking for gcc... gcc
> checking whether we are using the GNU C compiler... yes
> checking whether gcc accepts -g... yes
> checking for gcc option to accept ISO C89... none needed
> checking how to run the C preprocessor... gcc -E
> checking for h5cc... no
> checking for HDF5 type... neither
> checking for HDF5 libraries... no
> configure: WARNING:
> Unable to locate serial HDF5 compilation helper script 'h5cc'.
> Please specify --with-hdf5=<LOCATION> as the full path to h5cc.
> HDF5 support is being disabled (equivalent to --with-hdf5=no).
> 
> configure: Serial HDF5 support disabled
> configure: Parallel HDF5 support disabled
> 
> Supported file formats: netCDF
> 
> checking for libblas... no
> configure: error: "LAPACK requested but not found."
> 

config.log shows that the extern "C" declaration is throwing the error:

>  configure:7834:
>  checking for libblas
> configure:7856: gcc -o conftest -g -O2  -I/fusion/usc/opt/fftw/fftw-3.3.6_mpich-3.2_gcc-4.9.2/include -L/fusion/usc/opt/fftw/fftw-3.3.6_mpich-3.2_gcc-4.9.2/lib  conftes
> conftest.c:31:8: error: expected identifier or '(' before string constant
>  extern "C"
>         ^    
> configure:7856: $? = 1
> configure: failed program was: 
> | /* confdefs.h */
> | #define PACKAGE_NAME "BOUT++"
> | #define PACKAGE_TARNAME "bout--"
> | #define PACKAGE_VERSION "4.0.0"
> | #define PACKAGE_STRING "BOUT++ 4.0.0"
> | #define PACKAGE_BUGREPORT "bd512@york.ac.uk"
> | #define PACKAGE_URL ""
> | #define HAVE_MPI 1
> | #define HAVE_CXX11 1
> | #define HAVE_LIBM 1
> | #define STDC_HEADERS 1
> | #define HAVE_SYS_TYPES_H 1
> | #define HAVE_SYS_STAT_H 1
> | #define HAVE_STDLIB_H 1
> | #define HAVE_STRING_H 1
> | #define HAVE_MEMORY_H 1
> | #define HAVE_STRINGS_H 1
> | #define HAVE_INTTYPES_H 1
> | #define HAVE_STDINT_H 1
> | #define HAVE_UNISTD_H 1
> | #define HAVE_MALLOC_H 1
> | #define HAVE_STDLIB_H 1
> | #define HAVE_STRING_H 1
> | #define HAVE_STRINGS_H 1
> | #define HAVE_STDLIB_H 1
> | #define HAVE_MALLOC 1
> | #define HAVE_STDLIB_H 1
> | #define HAVE_REALLOC 1
> | #define HAVE_VPRINTF 1
> | /* end confdefs.h.  */
> | extern "C"
> |     char zgemm_ ();
> | int
> | main ()
> | {
> | return zgemm_ ();
> |   ;
> |   return 0;
> | }


Yet if you force configure to use C++ it continues successfully 
`./configure CC=g++ --with-lapack=/usr/lib64
`
Now configure shows:

> configure: Serial HDF5 support disabled
> configure: Parallel HDF5 support disabled
> 
> Supported file formats: netCDF
> 
> checking for libblas... yes
> checking for liblapack... yes
> Using LAPACK

I suggest fixing this by having configure use CXX to compile conftest.c rather than CC.

--- a/configure
+++ b/configure
@@ -5172,6 +5172,12 @@ fi
 
     { $as_echo "$as_me:${as_lineno-$LINENO}: checking for libfftw3" >&5
 $as_echo_n "checking for libfftw3... " >&6; }
+    ac_ext=cpp
+ac_cpp='$CXXCPP $CPPFLAGS'
+ac_compile='$CXX -c $CXXFLAGS $CPPFLAGS conftest.$ac_ext >&5'
+ac_link='$CXX -o conftest$ac_exeext $CXXFLAGS $CPPFLAGS $LDFLAGS conftest.$ac_ext $LIBS >&5'
+ac_compiler_gnu=$ac_cv_cxx_compiler_gnu
+
     LDFLAGS_save=$LDFLAGS
     LIBS="$EXTRA_LIBS -lfftw3"
     BACL_found=no
@@ -5249,6 +5255,12 @@ $as_echo "no" >&6; }
 else
   as_fn_error $? "\"FFTW3 is needed by BOUT++\"" "$LINENO" 5
 fi
+    ac_ext=cpp
+ac_cpp='$CXXCPP $CPPFLAGS'
+ac_compile='$CXX -c $CXXFLAGS $CPPFLAGS conftest.$ac_ext >&5'
+ac_link='$CXX -o conftest$ac_exeext $CXXFLAGS $CPPFLAGS $LDFLAGS conftest.$ac_ext $LIBS >&5'
+ac_compiler_gnu=$ac_cv_cxx_compiler_gnu
+
 
 
 echo ""
@@ -5498,6 +5510,12 @@ $as_echo "no" >&6; }
 
     { $as_echo "$as_me:${as_lineno-$LINENO}: checking for libnetcdf" >&5
 $as_echo_n "checking for libnetcdf... " >&6; }
+    ac_ext=cpp
+ac_cpp='$CXXCPP $CPPFLAGS'
+ac_compile='$CXX -c $CXXFLAGS $CPPFLAGS conftest.$ac_ext >&5'
+ac_link='$CXX -o conftest$ac_exeext $CXXFLAGS $CPPFLAGS $LDFLAGS conftest.$ac_ext $LIBS >&5'
+ac_compiler_gnu=$ac_cv_cxx_compiler_gnu
+
     LDFLAGS_save=$LDFLAGS
     LIBS="$EXTRA_LIBS -lnetcdf"
     BACL_found=no
@@ -5574,6 +5592,12 @@ $as_echo "no" >&6; }
 
     { $as_echo "$as_me:${as_lineno-$LINENO}: checking for libnetcdf_c++" >&5
 $as_echo_n "checking for libnetcdf_c++... " >&6; }
+    ac_ext=cpp
+ac_cpp='$CXXCPP $CPPFLAGS'
+ac_compile='$CXX -c $CXXFLAGS $CPPFLAGS conftest.$ac_ext >&5'
+ac_link='$CXX -o conftest$ac_exeext $CXXFLAGS $CPPFLAGS $LDFLAGS conftest.$ac_ext $LIBS >&5'
+ac_compiler_gnu=$ac_cv_cxx_compiler_gnu
+
     LDFLAGS_save=$LDFLAGS
     LIBS="$EXTRA_LIBS -lnetcdf_c++"
     BACL_found=no
@@ -5651,10 +5675,22 @@ $as_echo "no" >&6; }
 else
   NCFOUND=no
 fi
+    ac_ext=cpp
+ac_cpp='$CXXCPP $CPPFLAGS'
+ac_compile='$CXX -c $CXXFLAGS $CPPFLAGS conftest.$ac_ext >&5'
+ac_link='$CXX -o conftest$ac_exeext $CXXFLAGS $CPPFLAGS $LDFLAGS conftest.$ac_ext $LIBS >&5'
+ac_compiler_gnu=$ac_cv_cxx_compiler_gnu
+
 
 else
   NCFOUND=no
 fi
+    ac_ext=cpp
+ac_cpp='$CXXCPP $CPPFLAGS'
+ac_compile='$CXX -c $CXXFLAGS $CPPFLAGS conftest.$ac_ext >&5'
+ac_link='$CXX -o conftest$ac_exeext $CXXFLAGS $CPPFLAGS $LDFLAGS conftest.$ac_ext $LIBS >&5'
+ac_compiler_gnu=$ac_cv_cxx_compiler_gnu
+
 
 else
   NCFOUND=no
@@ -7316,6 +7352,12 @@ fi
 
     { $as_echo "$as_me:${as_lineno-$LINENO}: checking for libblas" >&5
 $as_echo_n "checking for libblas... " >&6; }
+    ac_ext=cpp
+ac_cpp='$CXXCPP $CPPFLAGS'
+ac_compile='$CXX -c $CXXFLAGS $CPPFLAGS conftest.$ac_ext >&5'
+ac_link='$CXX -o conftest$ac_exeext $CXXFLAGS $CPPFLAGS $LDFLAGS conftest.$ac_ext $LIBS >&5'
+ac_compiler_gnu=$ac_cv_cxx_compiler_gnu
+
     LDFLAGS_save=$LDFLAGS
     LIBS="$EXTRA_LIBS -lblas"
     BACL_found=no
@@ -7393,6 +7435,12 @@ $as_echo "no" >&6; }
 
     { $as_echo "$as_me:${as_lineno-$LINENO}: checking for liblapack" >&5
 $as_echo_n "checking for liblapack... " >&6; }
+    ac_ext=cpp
+ac_cpp='$CXXCPP $CPPFLAGS'
+ac_compile='$CXX -c $CXXFLAGS $CPPFLAGS conftest.$ac_ext >&5'
+ac_link='$CXX -o conftest$ac_exeext $CXXFLAGS $CPPFLAGS $LDFLAGS conftest.$ac_ext $LIBS >&5'
+ac_compiler_gnu=$ac_cv_cxx_compiler_gnu
+
     LDFLAGS_save=$LDFLAGS
     LIBS="$EXTRA_LIBS -llapack"
     BACL_found=no
@@ -7475,6 +7523,12 @@ else
   as_fn_error $? "\"LAPACK requested but not found.\"" "$LINENO" 5
 fi
 fi
+    ac_ext=cpp
+ac_cpp='$CXXCPP $CPPFLAGS'
+ac_compile='$CXX -c $CXXFLAGS $CPPFLAGS conftest.$ac_ext >&5'
+ac_link='$CXX -o conftest$ac_exeext $CXXFLAGS $CPPFLAGS $LDFLAGS conftest.$ac_ext $LIBS >&5'
+ac_compiler_gnu=$ac_cv_cxx_compiler_gnu
+
 
 
 else
@@ -7482,6 +7536,12 @@ else
   as_fn_error $? "\"LAPACK requested but not found.\"" "$LINENO" 5
 fi
 fi
+    ac_ext=cpp
+ac_cpp='$CXXCPP $CPPFLAGS'
+ac_compile='$CXX -c $CXXFLAGS $CPPFLAGS conftest.$ac_ext >&5'
+ac_link='$CXX -o conftest$ac_exeext $CXXFLAGS $CPPFLAGS $LDFLAGS conftest.$ac_ext $LIBS >&5'
+ac_compiler_gnu=$ac_cv_cxx_compiler_gnu
+
 
 
     echo ""

--- a/m4/bout.m4
+++ b/m4/bout.m4
@@ -36,6 +36,7 @@ dnl forth  arg is to be executed if not found
 dnl fifth  arg is an additional path to check
 AC_DEFUN([BOUT_ADDPATH_CHECK_LIB],[
     AC_MSG_CHECKING([for lib$1])
+    AC_LANG_PUSH([C++])
     LDFLAGS_save=$LDFLAGS
     LIBS="$EXTRA_LIBS -l$1"
     BACL_found=no
@@ -68,6 +69,7 @@ AC_DEFUN([BOUT_ADDPATH_CHECK_LIB],[
         AC_MSG_RESULT(no)
     fi
     AS_IF([test .$BACL_found = .yes], [$3],[$4])
+    AC_LANG_POP([C++])
 ])
 
 

